src/query/functions/src/aggregates/aggregate_histogram.rs

```diff
@@ -0,0 +1,615 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::collections::btree_map::BTreeMap;
+use std::collections::btree_map::Entry;
+use std::ops::AddAssign;
+use std::sync::Arc;
+
+use borsh::BorshDeserialize;
+use borsh::BorshSerialize;
+use databend_common_exception::ErrorCode;
+use databend_common_exception::Result;
+use databend_common_expression::types::decimal::*;
+use databend_common_expression::types::number::*;
+use databend_common_expression::types::string::StringColumnBuilder;
+use databend_common_expression::types::*;
+use databend_common_expression::with_number_mapped_type;
+use databend_common_expression::AggregateFunctionRef;
+use databend_common_expression::Scalar;
+use serde::Deserialize;
+use serde::Serialize;
+
+use super::FunctionData;
+use crate::aggregates::aggregate_function_factory::AggregateFunctionDescription;
+use crate::aggregates::aggregate_unary::UnaryState;
+use crate::aggregates::assert_variadic_arguments;
+use crate::aggregates::AggregateUnaryFunction;
+
+struct HistogramData {
+    pub max_num_buckets: u64,
+}
+
+impl FunctionData for HistogramData {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct HistogramState<T>
+where
+    T: ValueType,
+    T::Scalar: Ord + BorshSerialize + BorshDeserialize,
+{
+    pub value_map: BTreeMap<T::Scalar, u64>,
+}
+
+impl<T> Default for HistogramState<T>
+where
+    T: ValueType,
+    T::Scalar: Ord + BorshSerialize + BorshDeserialize,
+{
+    fn default() -> Self {
+        HistogramState::<T> {
+            value_map: BTreeMap::new(),
+        }
+    }
+}
+
+impl<T> UnaryState<T, StringType> for HistogramState<T>
+where
+    T: ValueType + Sync + Send,
+    T::Scalar: Ord + BorshSerialize + BorshDeserialize + Serialize + Sync + Send,
+{
+    fn add(&mut self, other: T::ScalarRef<'_>) -> Result<()> {
+        let other = T::to_owned_scalar(other);
```

Comment:
Is there any way to save a `to_vec` for `&[u8]` ? 

Maybe we should try use `get_mut`

Reply:
Me too, let's ignore this.